File: recursion/python/solutions/compact.py
"""Solution for compact."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def compact(lst):
    """
    Remove all None values.

    compact(1 -> None -> 2 -> None -> 3) => 1 -> 2 -> 3
    compact(()) => ()
    """
    if is_empty(lst):
        return EMPTY_LIST

    first, rest = unprepend(lst)

    if first is None:
        return compact(rest)
    else:
        return prepend(first, compact(rest))


if __name__ == '__main__':
    # Test cases
    assert to_python_list(compact(Node.from_list([1, None, 2, None, 3]))) == [1, 2, 3]
    assert to_python_list(compact(EMPTY_LIST)) == []
    assert to_python_list(compact(Node.from_list([None, None, None]))) == []
    assert to_python_list(compact(Node.from_list([1, 2, 3]))) == [1, 2, 3]

    print("All tests passed!")

File: recursion/javascript/solutions/compact.js
/**
 * Solution for compact.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function compact(list) {
  if (isEmpty(list)) {
    return EMPTY_LIST;
  }

  let [first, rest] = unprepend(list);

  if (first == null) {
    return compact(rest);
  }
  return prepend(first, compact(rest));
}

if (require.main === module) {
  console.assert(JSON.stringify(toArray(compact(Node.fromArray([1, null, 2, undefined, 3])))) === '[1,2,3]');
  console.assert(JSON.stringify(toArray(compact(Node.fromArray([null, null])))) === '[]');
  console.assert(JSON.stringify(toArray(compact(Node.fromArray([1, 2, 3])))) === '[1,2,3]');
  console.assert(JSON.stringify(toArray(compact(EMPTY_LIST))) === '[]');

  console.log('All tests passed!');
}

module.exports = { compact };
